### PR TITLE
Decl impl2

### PR DESCRIPTION
**Pull request by pbrunet**

### CODE REVIEW

**Comment by pbrunet** [comment]
For review most commits are just "move" but from b2da3fd, I cherry-pick part of the numpy.dot PR.
I added commit  a9f2e90 to valid it on travis and change include directive generation in 1205b80

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+SpaceAfterCStyleCast: false
+SpacesInContainerLiterals: true
+SpaceBeforeAssignmentOperators: true
+ContinuationIndentWidth: 4
+CommentPragmas:  '^ IWYU pragma:'
+ForEachMacros:   [ foreach, Q_FOREACH, BOOST_FOREACH ]
+SpaceBeforeParens: ControlStatements
+DisableFormat:   false
+...
+
```
Comment (serge-sans-paille) [inline .clang-format]: why not using one of the default setting?

**Comment by serge-sans-paille** [comment]
Did you rebase this on against master?

File: .clang-format (hunk 1 above)
Comment (pbrunet) [inline .clang-format]: It is the default I think. I dumped it so that it is stable for the future and easier to edit for our needs.

**Comment by pbrunet** [comment]
Yes it is.

File: .clang-format (hunk 1 above)
Comment (pbrunet) [inline .clang-format]: BTW, there are multiple default with clang-format. It could be: LLVM, google, chromium, mozilla or webkit. I choose LLVM I think.

**Pull request merged**